File: src/main/scala/crusader728/leetcode/dfs/NestedListWeightSumII364.scala
package scala.crusader728.leetcode.dfs

import scala.crusader728.leetcode.NestedInteger

object NestedListWeightSumII364 {
  def toMap(baseHeight: Int, ni: NestedInteger): Map[Int, List[Int]] = {
    if(ni.isInteger) {
      Map(baseHeight -> List(ni.getInteger))
    } else {
      ni.getList.map(elem => toMap(baseHeight + 1, elem))
        .foldLeft(Map.empty[Int, List[Int]]) {case (acc, sub) => {
          sub.foldLeft(acc) { case (m, (k, v)) => {
            m + (k -> (v ++ m.getOrElse(k, List.empty)))
          }}
        }}
    }
  }

  def depthSumInverse(nestedList: List[NestedInteger]): Int = {
    val weightMap = nestedList.map(ni => toMap(1, ni))
      .foldLeft(Map.empty[Int, List[Int]]) {case (acc, sub) => {
        sub.foldLeft(acc) {case (m, (k, v)) => {
          m + (k -> (v ++ m.getOrElse(k, List.empty)))
        }}
      }}
      .view
      .mapValues(l => l.sum)
    weightMap.keySet.maxOption match {
      case None => 0
      case Some(maxWeight) => {
        (1 to maxWeight).foldLeft(0) {case (acc, w) => {
          if(weightMap.contains(w)) {
            acc + (maxWeight - w + 1) * weightMap(w)
          } else {
            acc
          }
        }}
      }
    }

  }

}

File: src/main/withJava/crusader728/leetcode/dfs/NestedListWeightSumII364.java
package withJava.crusader728.leetcode.dfs;

import java.util.*;

public class NestedListWeightSumII364 {
    private interface NestedInteger {
        boolean isInteger();
        Integer getInteger();
        List<NestedInteger> getList();
    }

    public int depthSumInverse(List<NestedInteger> nestedList) {
        int[] sum = new int[1];
        int[] weightedSum = new int[1];
        int maxDepth = Integer.MIN_VALUE;
        for(NestedInteger nestedInteger: nestedList) {
            maxDepth = Math.max(maxDepth, dfs(nestedInteger, 1, weightedSum, sum));
        }
        return (maxDepth + 1) * sum[0] - weightedSum[0];
    }

    private int dfs(NestedInteger nestedInteger, int i, int[] weightedSum, int[] sum) {
        if(nestedInteger.isInteger()) {
            weightedSum[0] += i * nestedInteger.getInteger();
            sum[0] += nestedInteger.getInteger();
            return i;
        } else {
            int maxDepth = Integer.MIN_VALUE;
            for(NestedInteger ni: nestedInteger.getList()) {
                maxDepth = Math.max(maxDepth, dfs(ni, i + 1, weightedSum, sum));
            }
            return maxDepth;
        }
    }

}
